File: src/scpn_mif_core/physics/faraday_recovery.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN-MIF-CORE — MIF-009 Faraday induction recovery.
#
# OWNED-BY: scpn-mif-core
# CONSUMED-BY: scpn-mif-core
# SYNC-STATE: upstream-pending
# UPSTREAM-PIN: scpn-mif-core@0.0.1
# CONTRACT-TEST: tests/unit/physics/test_faraday_recovery.py
# TRACKED-ISSUE: docs/internal/upstream_contracts/04_scpn_fusion_core.md#c7-p1-post-poc-faraday-induction-back-emf-model
# LAST-SYNCED: 2026-06-04T0000
"""Faraday-law direct-recovery carrier for MIF-009.

The model computes the back-EMF induced in an effective recovery winding
by the time derivative of the external magnetic flux through the FRC
separatrix cross-section:

.. math::

    \\Phi(t) = B_\\mathrm{ext}(t) \\, \\pi R_s(t)^2,
    \\qquad
    \\mathcal{E}(t) = -N_\\mathrm{eff} \\, \\frac{d\\Phi}{dt}.

Applying the product rule gives the exact pointwise carrier used by the
Python, Rust, and Julia implementations:

.. math::

    \\frac{d\\Phi}{dt}
    = \\pi \\left(R_s^2 \\, \\dot B_\\mathrm{ext}
      + 2 R_s \\, \\dot R_s \\, B_\\mathrm{ext}\\right).

Status
------
``SYNC-STATE: upstream-pending`` — this is the local MIF-CORE exact
Faraday-law carrier. Self-consistent pulsed-compression coupling remains
owned by SCPN-FUSION-CORE (``FUS-C.6``); the eventual FUSION-hosted
recovery model is tracked as ``FUS-C.7``.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np
from numpy.typing import ArrayLike, NDArray

FloatArray = NDArray[np.float64]


@dataclass(frozen=True)
class FaradayRecoverySpec:
    """Immutable recovery-coil and load specification.

    Attributes
    ----------
    turns : float
        Positive effective turn count of the recovery coil. Fractional
        values are accepted to represent winding/coupling calibration.
    load_resistance_ohm : float
        Positive ohmic load presented to the induced EMF.
    coupling_efficiency : float
        Dimensionless power-transfer efficiency in ``[0, 1]``.
    """

    turns: float
    load_resistance_ohm: float
    coupling_efficiency: float = 1.0

    def __post_init__(self) -> None:
        turns = _require_finite("turns", self.turns)
        load = _require_finite("load_resistance_ohm", self.load_resistance_ohm)
        efficiency = _require_finite("coupling_efficiency", self.coupling_efficiency)
        if turns <= 0.0:
            raise ValueError("turns must be strictly positive")
        if load <= 0.0:
            raise ValueError("load_resistance_ohm must be strictly positive")
        if not 0.0 <= efficiency <= 1.0:
            raise ValueError("coupling_efficiency must lie in [0, 1]")
        object.__setattr__(self, "turns", turns)
        object.__setattr__(self, "load_resistance_ohm", load)
        object.__setattr__(self, "coupling_efficiency", efficiency)


@dataclass(frozen=True)
class FaradayRecoveryState:
    """Pointwise Faraday recovery observables."""

    radius_m: float
    radial_velocity_m_s: float
    magnetic_field_T: float
    magnetic_field_rate_T_s: float
    flux_Wb: float
    flux_rate_Wb_s: float
    back_emf_V: float
    recovered_power_W: float


@dataclass(frozen=True)
class FaradayRecoveryReport:
    """Waveform-level Faraday recovery observables."""

    time_s: FloatArray
    flux_Wb: FloatArray
    flux_rate_Wb_s: FloatArray
    back_emf_V: FloatArray
    recovered_power_W: FloatArray
    recovered_energy_J: float
    peak_abs_back_emf_V: float
    peak_recovered_power_W: float


def magnetic_flux(radius_m: float, magnetic_field_T: float) -> float:
    """Return external magnetic flux ``B_ext * pi * R_s**2`` in webers."""
    radius = _validate_radius(radius_m)
    field = _require_finite("magnetic_field_T", magnetic_field_T)
    return field * math.pi * radius * radius


def flux_rate(
    radius_m: float,
    radial_velocity_m_s: float,
    magnetic_field_T: float,
    magnetic_field_rate_T_s: float,
) -> float:
    """Return ``d(B_ext * pi * R_s**2) / dt`` in webers per second."""
    radius = _validate_radius(radius_m)
    velocity = _require_finite("radial_velocity_m_s", radial_velocity_m_s)
    field = _require_finite("magnetic_field_T", magnetic_field_T)
    field_rate = _require_finite("magnetic_field_rate_T_s", magnetic_field_rate_T_s)
    return math.pi * (radius * radius * field_rate + 2.0 * radius * velocity * field)


def faraday_back_emf(
    radius_m: float,
    radial_velocity_m_s: float,
    magnetic_field_T: float,
    magnetic_field_rate_T_s: float,
    turns: float,
) -> float:
    """Return induced back-EMF ``-turns * dPhi/dt`` in volts."""
    effective_turns = _require_finite("turns", turns)
    if effective_turns <= 0.0:
        raise ValueError("turns must be strictly positive")
    return -effective_turns * flux_rate(
        radius_m,
        radial_velocity_m_s,
        magnetic_field_T,
        magnetic_field_rate_T_s,
    )


def recovered_power(spec: FaradayRecoverySpec, back_emf_V: float) -> float:
    """Return instantaneous load power from a Thevenin EMF source."""
    emf = _require_finite("back_emf_V", back_emf_V)
    return spec.coupling_efficiency * emf * emf / spec.load_resistance_ohm


def evaluate_faraday_state(
    spec: FaradayRecoverySpec,
    radius_m: float,
    radial_velocity_m_s: float,
    magnetic_field_T: float,
    magnetic_field_rate_T_s: float,
) -> FaradayRecoveryState:
    """Evaluate pointwise flux, EMF, and recovered power."""
    flux = magnetic_flux(radius_m, magnetic_field_T)
    dflux_dt = flux_rate(
        radius_m,
        radial_velocity_m_s,
        magnetic_field_T,
        magnetic_field_rate_T_s,
    )
    emf = -spec.turns * dflux_dt
    power = recovered_power(spec, emf)
    return FaradayRecoveryState(
        radius_m=_validate_radius(radius_m),
        radial_velocity_m_s=_require_finite("radial_velocity_m_s", radial_velocity_m_s),
        magnetic_field_T=_require_finite("magnetic_field_T", magnetic_field_T),
        magnetic_field_rate_T_s=_require_finite("magnetic_field_rate_T_s", magnetic_field_rate_T_s),
        flux_Wb=flux,
        flux_rate_Wb_s=dflux_dt,
        back_emf_V=emf,
        recovered_power_W=power,
    )


def evaluate_faraday_recovery(
    spec: FaradayRecoverySpec,
    time_s: ArrayLike,
    radius_m: ArrayLike,
    radial_velocity_m_s: ArrayLike,
    magnetic_field_T: ArrayLike,
    magnetic_field_rate_T_s: ArrayLike,
) -> FaradayRecoveryReport:
    """Evaluate a full Faraday recovery waveform and integrate recovered energy.

    The velocity and field-rate arrays are explicit inputs so the waveform
    path uses the same exact product-rule carrier as the scalar path; no
    hidden finite-difference derivative is introduced here.
    """
    time = _as_1d_float_array("time_s", time_s)
    radii = _as_1d_float_array("radius_m", radius_m)
    velocities = _as_1d_float_array("radial_velocity_m_s", radial_velocity_m_s)
    fields = _as_1d_float_array("magnetic_field_T", magnetic_field_T)
    field_rates = _as_1d_float_array("magnetic_field_rate_T_s", magnetic_field_rate_T_s)
    _validate_same_shape(time, radii, velocities, fields, field_rates)
    if time.size < 2:
        raise ValueError("time_s must contain at least two samples")
    if not bool(np.all(np.diff(time) > 0.0)):
        raise ValueError("time_s must be strictly increasing")
    if not bool(np.all(radii >= 0.0)):
        raise ValueError("radius_m must be non-negative")

    flux = math.pi * radii * radii * fields
    dflux_dt = math.pi * (radii * radii * field_rates + 2.0 * radii * velocities * fields)
    emf = -spec.turns * dflux_dt
    power = spec.coupling_efficiency * emf * emf / spec.load_resistance_ohm
    dt = np.diff(time)
    energy = float(np.sum(0.5 * (power[:-1] + power[1:]) * dt))
    return FaradayRecoveryReport(
        time_s=_readonly(time),
        flux_Wb=_readonly(flux),
        flux_rate_Wb_s=_readonly(dflux_dt),
        back_emf_V=_readonly(emf),
        recovered_power_W=_readonly(power),
        recovered_energy_J=energy,
        peak_abs_back_emf_V=float(np.max(np.abs(emf))),
        peak_recovered_power_W=float(np.max(power)),
    )


def _require_finite(name: str, value: float) -> float:
    numeric = float(value)
    if not math.isfinite(numeric):
        raise ValueError(f"{name} must be finite")
    return numeric


def _validate_radius(radius_m: float) -> float:
    radius = _require_finite("radius_m", radius_m)
    if radius < 0.0:
        raise ValueError("radius_m must be non-negative")
    return radius


def _as_1d_float_array(name: str, values: ArrayLike) -> FloatArray:
    arr = np.asarray(values, dtype=np.float64)
    if arr.ndim != 1:
        raise ValueError(f"{name} must be a one-dimensional array")
    if arr.size == 0:
        raise ValueError(f"{name} must not be empty")
    if not bool(np.all(np.isfinite(arr))):
        raise ValueError(f"{name} must contain only finite values")
    return arr


def _validate_same_shape(reference: FloatArray, *arrays: FloatArray) -> None:
    for arr in arrays:
        if arr.shape != reference.shape:
            raise ValueError(
                "time_s, radius_m, radial_velocity_m_s, magnetic_field_T, and magnetic_field_rate_T_s must have the same shape"
            )


def _readonly(arr: FloatArray) -> FloatArray:
    copied = np.array(arr, dtype=np.float64, copy=True)
    copied.setflags(write=False)
    return copied

File: scpn-mif-rs/crates/mif-core/src/faraday_recovery.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SCPN-MIF-CORE — MIF-009 Faraday induction recovery.
//
// OWNED-BY: scpn-mif-core
// CONSUMED-BY: scpn-mif-core
// SYNC-STATE: upstream-pending
// UPSTREAM-PIN: scpn-mif-core@0.0.1
// CONTRACT-TEST: tests/unit/physics/test_faraday_recovery_rust_parity.py
// TRACKED-ISSUE: docs/internal/upstream_contracts/04_scpn_fusion_core.md#c7-p1-post-poc-faraday-induction-back-emf-model
// LAST-SYNCED: 2026-06-04T0000
//!
//! Exact Faraday-law direct-recovery carrier:
//! `EMF = -N_eff * d(B_ext * pi * R_s^2)/dt`.

use std::f64::consts::PI;

use thiserror::Error;

/// Immutable recovery-coil and load specification.
#[derive(Debug, Clone, Copy)]
pub struct FaradayRecoverySpec {
    /// Positive effective turn count.
    pub turns: f64,
    /// Positive ohmic load resistance.
    pub load_resistance_ohm: f64,
    /// Dimensionless transfer efficiency in `[0, 1]`.
    pub coupling_efficiency: f64,
}

impl FaradayRecoverySpec {
    /// Construct a validated recovery specification.
    pub fn new(
        turns: f64,
        load_resistance_ohm: f64,
        coupling_efficiency: f64,
    ) -> Result<Self, FaradayRecoveryError> {
        validate_positive("turns", turns)?;
        validate_positive("load_resistance_ohm", load_resistance_ohm)?;
        validate_finite("coupling_efficiency", coupling_efficiency)?;
        if !(0.0..=1.0).contains(&coupling_efficiency) {
            return Err(FaradayRecoveryError::CouplingEfficiencyOutOfRange);
        }
        Ok(Self {
            turns,
            load_resistance_ohm,
            coupling_efficiency,
        })
    }
}

/// Pointwise Faraday recovery observables.
#[derive(Debug, Clone, Copy)]
pub struct FaradayRecoveryState {
    /// Separatrix radius in metres.
    pub radius_m: f64,
    /// Separatrix radial velocity in metres per second.
    pub radial_velocity_m_s: f64,
    /// External axial magnetic field in tesla.
    pub magnetic_field_t: f64,
    /// External axial magnetic-field rate in tesla per second.
    pub magnetic_field_rate_t_s: f64,
    /// Magnetic flux in webers.
    pub flux_wb: f64,
    /// Magnetic flux rate in webers per second.
    pub flux_rate_wb_s: f64,
    /// Induced back-EMF in volts.
    pub back_emf_v: f64,
    /// Recovered load power in watts.
    pub recovered_power_w: f64,
}

/// Waveform-level Faraday recovery observables.
#[derive(Debug, Clone)]
pub struct FaradayRecoveryReport {
    /// Magnetic flux samples in webers.
    pub flux_wb: Vec<f64>,
    /// Magnetic flux-rate samples in webers per second.
    pub flux_rate_wb_s: Vec<f64>,
    /// Induced back-EMF samples in volts.
    pub back_emf_v: Vec<f64>,
    /// Recovered load-power samples in watts.
    pub recovered_power_w: Vec<f64>,
    /// Trapezoid-integrated recovered energy in joules.
    pub recovered_energy_j: f64,
    /// Peak absolute back-EMF in volts.
    pub peak_abs_back_emf_v: f64,
    /// Peak recovered power in watts.
    pub peak_recovered_power_w: f64,
}

/// Validation error for Faraday recovery inputs.
#[derive(Debug, Error, PartialEq, Eq)]
pub enum FaradayRecoveryError {
    /// A scalar field was not finite.
    #[error("{field} must be finite")]
    NonFinite {
        /// Field name.
        field: &'static str,
    },
    /// A scalar field was not strictly positive.
    #[error("{field} must be strictly positive")]
    NonPositive {
        /// Field name.
        field: &'static str,
    },
    /// Radius must be non-negative.
    #[error("radius_m must be non-negative")]
    NegativeRadius,
    /// Coupling efficiency must lie in the unit interval.
    #[error("coupling_efficiency must lie in [0, 1]")]
    CouplingEfficiencyOutOfRange,
    /// Waveform arrays must share the same length.
    #[error(
        "time_s, radius_m, radial_velocity_m_s, magnetic_field_t, and magnetic_field_rate_t_s must have the same length"
    )]
    LengthMismatch,
    /// Waveform arrays must contain at least two samples.
    #[error("time_s must contain at least two samples")]
    TooFewSamples,
    /// Time samples must be strictly increasing.
    #[error("time_s must be strictly increasing")]
    NonIncreasingTime,
    /// Radius samples must be non-negative.
    #[error("radius_m must be non-negative")]
    NegativeRadiusSample,
}

/// Return external magnetic flux `B_ext * pi * R_s^2` in webers.
pub fn magnetic_flux(radius_m: f64, magnetic_field_t: f64) -> Result<f64, FaradayRecoveryError> {
    let radius = validate_radius(radius_m)?;
    let field = validate_finite("magnetic_field_t", magnetic_field_t)?;
    Ok(field * PI * radius * radius)
}

/// Return `d(B_ext * pi * R_s^2) / dt` in webers per second.
pub fn flux_rate(
    radius_m: f64,
    radial_velocity_m_s: f64,
    magnetic_field_t: f64,
    magnetic_field_rate_t_s: f64,
) -> Result<f64, FaradayRecoveryError> {
    let radius = validate_radius(radius_m)?;
    let velocity = validate_finite("radial_velocity_m_s", radial_velocity_m_s)?;
    let field = validate_finite("magnetic_field_t", magnetic_field_t)?;
    let field_rate = validate_finite("magnetic_field_rate_t_s", magnetic_field_rate_t_s)?;
    Ok(PI * (radius * radius * field_rate + 2.0 * radius * velocity * field))
}

/// Return induced back-EMF `-turns * dPhi/dt` in volts.
pub fn faraday_back_emf(
    radius_m: f64,
    radial_velocity_m_s: f64,
    magnetic_field_t: f64,
    magnetic_field_rate_t_s: f64,
    turns: f64,
) -> Result<f64, FaradayRecoveryError> {
    let turns = validate_positive("turns", turns)?;
    Ok(-turns
        * flux_rate(
            radius_m,
            radial_velocity_m_s,
            magnetic_field_t,
            magnetic_field_rate_t_s,
        )?)
}

/// Return instantaneous recovered load power in watts.
pub fn recovered_power(
    spec: &FaradayRecoverySpec,
    back_emf_v: f64,
) -> Result<f64, FaradayRecoveryError> {
    let emf = validate_finite("back_emf_v", back_emf_v)?;
    Ok(spec.coupling_efficiency * emf * emf / spec.load_resistance_ohm)
}

/// Evaluate pointwise flux, EMF, and recovered power.
pub fn evaluate_faraday_state(
    spec: &FaradayRecoverySpec,
    radius_m: f64,
    radial_velocity_m_s: f64,
    magnetic_field_t: f64,
    magnetic_field_rate_t_s: f64,
) -> Result<FaradayRecoveryState, FaradayRecoveryError> {
    let flux_wb = magnetic_flux(radius_m, magnetic_field_t)?;
    let flux_rate_wb_s = flux_rate(
        radius_m,
        radial_velocity_m_s,
        magnetic_field_t,
        magnetic_field_rate_t_s,
    )?;
    let back_emf_v = -spec.turns * flux_rate_wb_s;
    let recovered_power_w = recovered_power(spec, back_emf_v)?;
    Ok(FaradayRecoveryState {
        radius_m: validate_radius(radius_m)?,
        radial_velocity_m_s: validate_finite("radial_velocity_m_s", radial_velocity_m_s)?,
        magnetic_field_t: validate_finite("magnetic_field_t", magnetic_field_t)?,
        magnetic_field_rate_t_s: validate_finite(
            "magnetic_field_rate_t_s",
            magnetic_field_rate_t_s,
        )?,
        flux_wb,
        flux_rate_wb_s,
        back_emf_v,
        recovered_power_w,
    })
}

/// Evaluate a full Faraday recovery waveform and integrate recovered energy.
pub fn evaluate_faraday_recovery(
    spec: &FaradayRecoverySpec,
    time_s: &[f64],
    radius_m: &[f64],
    radial_velocity_m_s: &[f64],
    magnetic_field_t: &[f64],
    magnetic_field_rate_t_s: &[f64],
) -> Result<FaradayRecoveryReport, FaradayRecoveryError> {
    validate_waveform_inputs(
        time_s,
        radius_m,
        radial_velocity_m_s,
        magnetic_field_t,
        magnetic_field_rate_t_s,
    )?;
    let mut flux_wb = Vec::with_capacity(time_s.len());
    let mut flux_rate_wb_s = Vec::with_capacity(time_s.len());
    let mut back_emf_v = Vec::with_capacity(time_s.len());
    let mut recovered_power_w = Vec::with_capacity(time_s.len());
    let mut peak_abs_back_emf_v = 0.0_f64;
    let mut peak_recovered_power_w = 0.0_f64;

    for idx in 0..time_s.len() {
        let flux = magnetic_flux(radius_m[idx], magnetic_field_t[idx])?;
        let rate = flux_rate(
            radius_m[idx],
            radial_velocity_m_s[idx],
            magnetic_field_t[idx],
            magnetic_field_rate_t_s[idx],
        )?;
        let emf = -spec.turns * rate;
        let power = recovered_power(spec, emf)?;
        flux_wb.push(flux);
        flux_rate_wb_s.push(rate);
        back_emf_v.push(emf);
        recovered_power_w.push(power);
        peak_abs_back_emf_v = peak_abs_back_emf_v.max(emf.abs());
        peak_recovered_power_w = peak_recovered_power_w.max(power);
    }

    let mut recovered_energy_j = 0.0;
    for idx in 0..time_s.len() - 1 {
        let dt = time_s[idx + 1] - time_s[idx];
        recovered_energy_j += 0.5 * (recovered_power_w[idx] + recovered_power_w[idx + 1]) * dt;
    }

    Ok(FaradayRecoveryReport {
        flux_wb,
        flux_rate_wb_s,
        back_emf_v,
        recovered_power_w,
        recovered_energy_j,
        peak_abs_back_emf_v,
        peak_recovered_power_w,
    })
}

fn validate_waveform_inputs(
    time_s: &[f64],
    radius_m: &[f64],
    radial_velocity_m_s: &[f64],
    magnetic_field_t: &[f64],
    magnetic_field_rate_t_s: &[f64],
) -> Result<(), FaradayRecoveryError> {
    let len = time_s.len();
    if len < 2 {
        return Err(FaradayRecoveryError::TooFewSamples);
    }
    if radius_m.len() != len
        || radial_velocity_m_s.len() != len
        || magnetic_field_t.len() != len
        || magnetic_field_rate_t_s.len() != len
    {
        return Err(FaradayRecoveryError::LengthMismatch);
    }
    for idx in 0..len {
        validate_finite("time_s", time_s[idx])?;
        validate_radius(radius_m[idx]).map_err(|_| FaradayRecoveryError::NegativeRadiusSample)?;
        validate_finite("radial_velocity_m_s", radial_velocity_m_s[idx])?;
        validate_finite("magnetic_field_t", magnetic_field_t[idx])?;
        validate_finite("magnetic_field_rate_t_s", magnetic_field_rate_t_s[idx])?;
        if idx > 0 && time_s[idx] <= time_s[idx - 1] {
            return Err(FaradayRecoveryError::NonIncreasingTime);
        }
    }
    Ok(())
}

fn validate_finite(field: &'static str, value: f64) -> Result<f64, FaradayRecoveryError> {
    if value.is_finite() {
        Ok(value)
    } else {
        Err(FaradayRecoveryError::NonFinite { field })
    }
}

fn validate_positive(field: &'static str, value: f64) -> Result<f64, FaradayRecoveryError> {
    validate_finite(field, value)?;
    if value > 0.0 {
        Ok(value)
    } else {
        Err(FaradayRecoveryError::NonPositive { field })
    }
}

fn validate_radius(radius_m: f64) -> Result<f64, FaradayRecoveryError> {
    validate_finite("radius_m", radius_m)?;
    if radius_m >= 0.0 {
        Ok(radius_m)
    } else {
        Err(FaradayRecoveryError::NegativeRadius)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn constant_field_expanding_radius_matches_closed_form() {
        let emf = faraday_back_emf(0.2, 800.0, 5.0, 0.0, 12.0).unwrap();
        let expected = -12.0 * (2.0 * PI * 0.2 * 800.0 * 5.0);
        assert!((emf - expected).abs() <= expected.abs() * 1e-15);
    }

    #[test]
    fn constant_radius_field_ramp_matches_closed_form() {
        let emf = faraday_back_emf(0.17, 0.0, 3.0, 25_000.0, 48.0).unwrap();
        let expected = -48.0 * (PI * 0.17_f64.powi(2) * 25_000.0);
        assert!((emf - expected).abs() <= expected.abs() * 1e-15);
    }

    #[test]
    fn static_radius_and_field_have_zero_emf() {
        assert_eq!(faraday_back_emf(0.4, 0.0, 8.0, 0.0, 32.0).unwrap(), 0.0);
    }

    #[test]
    fn waveform_integrates_constant_power_exactly() {
        let spec = FaradayRecoverySpec::new(20.0, 5.0, 0.8).unwrap();
        let time_s = [0.0, 0.5, 1.0, 1.5];
        let radius_m = [0.1, 0.1, 0.1, 0.1];
        let velocity = [0.0, 0.0, 0.0, 0.0];
        let field = [3.0, 4.0, 5.0, 6.0];
        let field_rate = [2.0, 2.0, 2.0, 2.0];
        let report =
            evaluate_faraday_recovery(&spec, &time_s, &radius_m, &velocity, &field, &field_rate)
                .unwrap();
        let expected_emf = -spec.turns * PI * 0.1_f64.powi(2) * 2.0;
        let expected_power =
            spec.coupling_efficiency * expected_emf.powi(2) / spec.load_resistance_ohm;
        assert!((report.back_emf_v[0] - expected_emf).abs() < 1e-15);
        assert!((report.recovered_energy_j - expected_power * 1.5).abs() < 1e-15);
    }

    #[test]
    fn rejects_non_positive_turns() {
        assert_eq!(
            FaradayRecoverySpec::new(0.0, 1.0, 1.0).unwrap_err(),
            FaradayRecoveryError::NonPositive { field: "turns" }
        );
    }

    #[test]
    fn rejects_non_increasing_time() {
        let spec = FaradayRecoverySpec::new(10.0, 1.0, 1.0).unwrap();
        let err = evaluate_faraday_recovery(
            &spec,
            &[0.0, 1.0, 1.0],
            &[0.1, 0.1, 0.1],
            &[0.0, 0.0, 0.0],
            &[1.0, 1.0, 1.0],
            &[0.0, 0.0, 0.0],
        )
        .unwrap_err();
        assert_eq!(err, FaradayRecoveryError::NonIncreasingTime);
    }
}
